Scripts/Core/EditorVR.Menus.cs

```diff
@@ -197,119 +109,282 @@ internal void UpdateMenuVisibilities()
+					var hideDistance = deviceData.mainMenu.hideDistance;
+					if (scaledPointerDistance < hideDistance + k_MenuHideMargin)
 					{
-						deviceData.menuHideFlags[menu] &= ~MenuHideFlags.OverActivator;
+						foreach (var menu in menus)
+						{
+							// Only set if hidden--value is reset every frame
+							deviceData.menuHideFlags[menu] |= MenuHideFlags.OverUI;
+						}
+
+						return true;
 					}
+
+					return (deviceData.menuHideFlags[deviceData.mainMenu] & MenuHideFlags.Hidden) != 0;
 				}
+
+				return true;
 			}
 
 			internal static void UpdateAlternateMenuOnSelectionChanged(Transform rayOrigin)
 			{
+				if (rayOrigin == null)
+					return;
+
 				SetAlternateMenuVisibility(rayOrigin, Selection.gameObjects.Length > 0);
 			}
 
 			internal static void SetAlternateMenuVisibility(Transform rayOrigin, bool visible)
 			{
 				Rays.ForEachProxyDevice(deviceData =>
 				{
+					var menuHideFlags = deviceData.menuHideFlags;
 					var alternateMenu = deviceData.alternateMenu;
 					if (alternateMenu != null)
 					{
-						var flags = deviceData.menuHideFlags[alternateMenu];
-						deviceData.menuHideFlags[alternateMenu] = (deviceData.rayOrigin == rayOrigin) && visible ? flags & ~MenuHideFlags.Hidden : flags | MenuHideFlags.Hidden;
+						var alternateMenuFlags = menuHideFlags[alternateMenu];
+						menuHideFlags[alternateMenu] = (deviceData.rayOrigin == rayOrigin) && visible ? alternateMenuFlags & ~MenuHideFlags.Hidden : alternateMenuFlags | MenuHideFlags.Hidden;
+
+						if ((menuHideFlags[alternateMenu] & MenuHideFlags.Hidden) != 0)
+						{
+							var customMenu = deviceData.customMenu;
+
+							if (customMenu != null && (menuHideFlags[deviceData.mainMenu] & MenuHideFlags.Hidden) != 0)
+							{
+								menuHideFlags[customMenu] &= ~MenuHideFlags.OtherMenu;
+							}
+						}
 					}
 				});
 			}
 
 			internal static void OnMainMenuActivatorSelected(Transform rayOrigin, Transform targetRayOrigin)
 			{
-				var deviceData = evr.m_DeviceData.FirstOrDefault(dd => dd.rayOrigin == rayOrigin);
-				if (deviceData != null)
+				if (evr.m_DeviceData.Any(deviceData => deviceData.rayOrigin == rayOrigin && deviceData.mainMenuActivator.disabled))
+					return;
+
+				foreach (var deviceData in evr.m_DeviceData)
 				{
 					var mainMenu = deviceData.mainMenu;
 					if (mainMenu != null)
 					{
+						var customMenu = deviceData.customMenu;
+						var alternateMenu = deviceData.alternateMenu;
 						var menuHideFlags = deviceData.menuHideFlags;
-						menuHideFlags[mainMenu] ^= MenuHideFlags.Hidden;
+						var alternateMenuVisible = alternateMenu != null && (menuHideFlags[alternateMenu] & MenuHideFlags.Hidden) == 0;
 
-						var customMenu = deviceData.customMenu;
-						if (customMenu != null)
-							menuHideFlags[customMenu] &= ~MenuHideFlags.Hidden;
+						deviceData.menuAutoShowTimes[mainMenu] = 0;
+
+						if (deviceData.rayOrigin == rayOrigin)
+						{
+							menuHideFlags[mainMenu] ^= MenuHideFlags.Hidden;
```

Comment:
yep. Should I comment? This is one of the few things that hasn't changed from the previous implementation :)